pex/specifier_sets.py

```diff
@@ -0,0 +1,331 @@
+    symbol = attr.ib(init=False, eq=False, repr=False)  # type: str
+    rank = attr.ib(init=False, eq=False, repr=False)  # type: int
+
+    def __attrs_post_init__(self):
+        raise NotImplementedError(
+            "_Bound must be subclassed and __attrs_post_init__ implemented to initialize symbol"
+            "and rank."
+        )
+
+    def __lt__(self, other):
+        # type: (Any) -> bool
+        if type(self) != type(other):
+            return NotImplemented
+        return (self.version, self.rank) < (other.version, other.rank)
+
+    def __str__(self):
+        # type: () -> str
+        parts = [self.symbol]
+        if self.inclusive:
+            parts.append("=")
+        parts.append(str(self.version))
+        return "".join(parts)
+
+
+class LowerBound(_Bound):
+    def __attrs_post_init__(self):
+        # type: () -> None
+        object.__setattr__(self, "symbol", ">")
+        # I.E sorted order of: >=1, >1, >=2, >2
+        object.__setattr__(self, "rank", 0 if self.inclusive else 1)
+
+
+class UpperBound(_Bound):
+    def __attrs_post_init__(self):
+        # type: () -> None
+        object.__setattr__(self, "symbol", "<")
+        # I.E sorted order of: <1, <=1, <2, <=2
+        object.__setattr__(self, "rank", 1 if self.inclusive else 0)
+
+
+@attr.s(frozen=True)
+class ExcludedRange(object):
+    lower = attr.ib()  # type: LowerBound
+    upper = attr.ib()  # type: UpperBound
+
+    def intersects(self, other):
+        # type: (Range) -> bool
+
+        # Test partial overlap from below.
+        if other.upper:
+            if self.lower.version > other.upper.version:
+                return False
+            elif (
+                self.lower.version == other.upper.version
+                and self.lower.inclusive
+                and other.upper.inclusive
+            ):
+                return True
+
+        # Test partial overlap from above.
+        if other.lower:
+            if self.upper.version < other.lower.version:
+                return False
+            elif (
+                self.upper.version == other.lower.version
+                and self.upper.inclusive
+                and other.lower.inclusive
+            ):
+                return True
+
+        # Test full containment.
+        excluded_range = Range(lower=self.lower, upper=self.upper)
+        return other in excluded_range or excluded_range in other
+
+    def __str__(self):
+        return ",".join(map(str, (self.lower, self.upper)))
+
+
+@attr.s(frozen=True)
+class Range(object):
+    lower = attr.ib(default=None)  # type: Optional[LowerBound]
+    upper = attr.ib(default=None)  # type: Optional[UpperBound]
+    excludes = attr.ib(default=())  # type: Tuple[ExcludedRange, ...]
+
+    def __contains__(self, other):
+        # type: (Any) -> bool
+        if Range is not type(other):
+            return NotImplemented
+        return self.contains(other)
+
+    def contains(self, other):
+        # type: (Range) -> bool
```

Comment:
If anything I'd bias opposite. In the bespoke functions (`contains` and `intersect`), I leverage type annotations and MyPy. As a matter of keeping my own sanity, I never double-check type-checking with runtime assertions. It may be the case that I could type `__contains__` to only accept other `Range`s, but that's against the spirit of the `__contains__` interface, which is supposed to work on `Any`.